src/bert_record_validator.erl

```diff
@@ -0,0 +1,216 @@
+    {Clause, []};
+get_type({integer, _}, Name, Var) ->
+    get_type(integer, Name, Var);
+get_type(integer, Name, Var) ->
+    Clause = ?QUOTE(({_@Name@, _@Var}) when is_integer(_@Var) -> []),
+    {Clause, []};
+get_type({non_neg_integer, _}, Name, Var) ->
+    Clause = ?QUOTE(({_@Name@, _@Var}) when is_integer(_@Var) andalso _@Var >= 0 -> []),
+    {Clause, []};
+get_type({pos_integer, _}, Name, Var) ->
+    Clause = ?QUOTE(({_@Name@, _@Var}) when is_integer(_@Var) andalso _@Var > 0 -> []),
+    {Clause, []};
+get_type({neg_integer, _}, Name, Var) ->
+    Clause = ?QUOTE(({_@Name@, _@Var}) when is_integer(_@Var) andalso _@Var < 0 -> []),
+    {Clause, []};
+get_type({boolean, _}, Name, Var) ->
+    Clause = ?QUOTE(({_@Name@, _@Var}) when is_boolean(_@Var) -> []),
+    {Clause, []};
+get_type({pid, _}, Name, Var) ->
+    Clause = ?QUOTE(({_@Name@, _@Var}) when is_pid(_@Var) -> []),
+    {Clause, []};
+get_type({node, _}, Name, Var) ->
+    Clause = ?QUOTE(({_@Name@, _@Var}) when is_atom(_@Var) -> []),
+    {Clause, []};
+get_type({list,[{type,_,record,[{atom,_,C}]}]}, Name, Var) ->
+    RecVar = erl_syntax:variable(lists:concat([ erl_syntax:variable_literal(Var), "_", C ])),
+    Clause = ?QUOTE(({_@Name@, _@Var}) when is_list(_@Var) ->
+                           [ {_@Name@, _@RecVar} || _@RecVar <- _@Var, not is_record(_@RecVar, _@C@)]),
+    {Clause, {Name, C}};  %% These are handled later
+get_type({list,[{type,_,union, Union}]}, Name, Var) when is_list(Union) ->
+    UVar = erl_syntax:variable(lists:concat([ erl_syntax:variable_literal(Var), "Element"])),
+    Clause =
+        case orelse_it([ type_validate(E, UVar) || E <- Union ]) of
+            [] ->
+                ?QUOTE(({_@Name@, _@Var}) when is_list(_@Var) -> []);
+            Guard ->
+                ?QUOTE(({_@Name@, _@Var}) when is_list(_@Var) ->
+                                       [ {_@Name@, _@UVar} || _@UVar <- _@Var, not (_@Guard) ])
+        end,
+    {Clause, []};
+get_type({list,_}, Name, Var) ->
+    Clause = ?QUOTE(({_@Name@, _@Var}) when is_list(_@Var) -> []),
+    {Clause, []};
+get_type({record,[{atom, _, Record}]}, Name, Var) ->
+    Clause = ?QUOTE(({_@Name@, _@Var}) when is_record(_@Var, _@Record@) -> []),
+    {Clause, {Name, Record}};
+get_type({term, []}, Name, Var) ->
+    Clause = ?QUOTE(({_@Name@, _}) -> []),
+    {Clause, []};
+get_type({tuple,_}, Name, Var) ->
+    Clause = ?QUOTE(({_@Name@, _@Var}) when is_tuple(_@Var) -> []),
+    {Clause, []};
+get_type({binary,_}, Name, Var) ->
+    Clause = ?QUOTE(({_@Name@, _@Var}) when is_binary(_@Var) -> []),
+    {Clause, []};
+get_type({union, Union}, Name, Var) when is_list(Union) ->
+    Clause =
+        case orelse_it([ type_validate(E, Var) || E <- Union ]) of
+            [] ->
+                ?QUOTE(({_@Name@, _}) -> []);
+            Guard ->
+                erl_syntax:clause([?QUOTE({_@Name@, _@Var})], Guard, [?QUOTE([])])
+        end,
+    {Clause, []};
+get_type(_Type, _Name, _Var) ->
```

Comment:
How would you notice that this has happened?

Reply:
We would not look at the source code and won't see that this case kicks in. That's the same presently, if there is a user defined type, no check will be performed and we are not warned. Not super powerful, but backward compatible.